File: 0859-buddy-strings/0859-buddy-strings.java
class Solution {
    public boolean buddyStrings(String s, String goal) {
        if(s.length() != goal.length()) {
            return false;
        }
        if(s.equals(goal)) {
            int[] freq = new int[26];
            for(char ch : s.toCharArray()) {
                freq[ch - 'a'] += 1;
                if(freq[ch - 'a'] == 2) {
                    return true;
                }
            }
            return false;
        }
        int fi = -1;
        int si = -1;
        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(i) != goal.charAt(i)) {
                if(fi == -1) {
                    fi = i;
                } else if(si == -1) {
                    si = i;
                } else {
                    return false;
                }
            }
        }
        if(si == -1) {
            return false;
        }
        return s.charAt(fi) == goal.charAt(si) && s.charAt(si) == goal.charAt(fi);
    }
}

File: 0859-buddy-strings/0859-buddy-strings.py
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal):
            return False
        if s == goal:
            freq = [0 for _ in range(26)]
            for c in s:
                freq[ord(c) - ord('a')] += 1
                if freq[ord(c) - ord('a')] == 2:
                    return True
        fi = -1
        si = -1
        for i in range(len(s)):
            if s[i] != goal[i]:
                if fi == -1:
                    fi = i
                elif si == -1:
                    si = i
                else:
                    return False
        if si == -1:
            return False
        return s[fi] == goal[si] and s[si] == goal[fi]
                